### PR TITLE
subcommands/bashstyle refactor

### LINKED ISSUE #1579: Plugin command structure
Currently, plugin command structure (i.e. anything that is not a trigger. think `dokku ps:restart`) lives in a `commands` file that must be "hunted for." 

execution today:
```
implemented=0
for script in $PLUGIN_ENABLED_PATH/*/commands; do
  set +e; $script "$@" ; exit_code=$? ; set -e
  if [[ "$exit_code" -eq "$DOKKU_NOT_IMPLEMENTED_EXIT" ]]; then
    continue
  fi

  implemented=1
  if [[ "$exit_code" -ne "$DOKKU_VALID_EXIT" ]]; then
    exit $exit_code
  fi
done
if [[ "$implemented" -eq 0 ]]; then
  dokku_log_warn "\`$*\` is not a dokku command."
  dokku_log_warn "See \`dokku help\` for a list of available commands."
  exit 1
fi    
```

I propose this be simplified in a similar manner as #1039 and #592. 

### benefits
- more efficient execution
- less TRACE output leads to easier debugging
- explicit plugin namespace prevents plugins from polluting other plugin namespaces. this also prevents an error in one plugin from blocking the execution of another

proposed execution example:
```
PLUGIN_NAME="$1"
if [[ -x $PLUGIN_ENABLED_PATH/$PLUGIN_NAME/command ]]; then
  "$PLUGIN_ENABLED_PATH/$PLUGIN_NAME/command" "$@"
elif [[ -x $PLUGIN_ENABLED_PATH/${PLUGIN_NAME%%:*}/subcommands/${1#*:} ]]; then
  "$PLUGIN_ENABLED_PATH/${PLUGIN_NAME%%:*}/subcommands/${1#*:}" "$@"
fi
```
refs: #592 #1039

---

### general spec
- each command lives in `plugins/<plugin_name>/subcommands/<name_of_command>`
- current UI will be maintained (i.e. `<plugin_name>:<subcommand>`)
- top-level command (i.e. `dokku ps`) will be implemented in `plugins/<plugin_name>/command`
- old method of "hunting" may be supported as an interim but will be deprecated
- execution of plugin triggers are not changed by this

### open questions
- Can we actually do this with all of the core plugins? probably not as `git` comes to mind
- If we still support the old method, do we need to wait for a major release to merge this?

thoughts?

### PR DESCRIPTION
**Pull request by michaelshobbs**
- [x] moves commands to there own files
- [x] implements `--app` refs #1638 
- [x] everything that can be a function is a function (i.e. `dokku` is not called and we have limited global vars defined
- [x] docs
- [x] tests (with `--app` support)
- [x] fixes client issue #1638

refs #592 #1039
closes #1579 #1638

### CODE REVIEW

File: docs/development/plugin-creation.md [hunk 1]
```diff
@@ -8,7 +8,89 @@ If you create your own plugin:
...
+
+  *)
+    exit $DOKKU_NOT_IMPLEMENTED_EXIT
+    ;;
+
+esac
+```
+
+
+### Sample plugin - old structure (still supported but not advised)
```
Comment (josegonzalez) [inline docs/development/plugin-creation.md]: Just remove the old method. One true way.

Comment (michaelshobbs) [inline docs/development/plugin-creation.md]: Sounds good to me

File: plugins/20_events/subcommands/default [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
+source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
+
+events_main_cmd() {
+  local desc="shows contents of dokku events log"
```
Comment (josegonzalez) [inline plugins/20_events/subcommands/default]: Worth moving these to declare? *shrug*

Comment (michaelshobbs) [inline plugins/20_events/subcommands/default]: Yeah I can. Just `declare desc=`, right?

File: plugins/20_events/subcommands/default [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
+source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
+
+events_main_cmd() {
```
Comment (josegonzalez) [inline plugins/20_events/subcommands/default]: Was hoping we'd just call all the functions `main`, but this is certainly useful if we ever wanted to start sourcing the subcommands.

File: plugins/20_events/subcommands/default (hunk 2 above)
Comment (josegonzalez) [inline plugins/20_events/subcommands/default]: Yep

File: plugins/20_events/subcommands/default (hunk 3 above)
Comment (michaelshobbs) [inline plugins/20_events/subcommands/default]: Yeah the thinking there was clarity in trace output

**Comment by josegonzalez** [comment]
:+1:

**Pull request merged**